File: leetcode/Array/0300_Longest_Increasing_Subsequence.cpp
// O(NlogN), binary search
class Solution {
public:
	int lengthOfLIS(vector<int>& nums) {
		if (nums.size() == 0)
			return 0;
		
		vector<int> IS;
		IS.push_back(nums[0]);
		for (int i = 1; i < nums.size(); i++) {
			// find nums[i] should be inserted in which position in increasing subsequence
			int l = 0, r = IS.size() - 1;
			while (l <= r) {
				int mid = (l + r) / 2;
				if (IS[mid] < nums[i])
					l = mid + 1;
				else
					r = mid - 1;
			}
			if (l < IS.size())
				IS[l] = nums[i];
			else
				IS.push_back(nums[i]);
		}
		return IS.size();
	}
};

// O(N^2), dynamic programming
class Solution {
public:
	int lengthOfLIS(vector<int>& nums) {
		if (nums.size() == 0)
			return 0;
		
		int n = nums.size(), maxLength = 1;
		vector<int> ISlength(n, 1);
		// ISlength[i]: the max length of increasing subsequence until nums[i]

		for (int i = 1; i < n; i++) {
			for (int j = i - 1; j >= 0; j--) {
				if (nums[i] > nums[j])
					ISlength[i] = max(ISlength[i], ISlength[j] + 1);
			}
			maxLength = max(maxLength, ISlength[i]);
		}
		return maxLength;
	}
};

File: leetcode/Binary_Search/0300_Longest_Increasing_Subsequence.py
# O(N^2), DP
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        N = len(nums)
        result = 1
        dp = [1] * N # dp[i]: the max len of increasing sequence until nums[i], including nums[i]

        for i in range(1, N):
            for j in range(i):
                if nums[i] > nums[j]: # can append nums[i] in increasing sequence ends at nums[j]
                    dp[i] = max(dp[i], dp[j] + 1)

            result = max(result, dp[i])

        return result

# O(NlogN), binary search
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        N = len(nums)
        IS = [nums[0]]

        for i in range(1, N):
            # find nums[i] should be inserted in which position in increasing subsequence
            left, right = 0, len(IS) - 1
            while left <= right:
                mid = (left + right) // 2
                if IS[mid] < nums[i]:
                    left = mid + 1
                else:
                    right = mid - 1
            # the final L is the index of first occurrence or index for inserting
            if left < len(IS):
                IS[left] = nums[i]
            else:
                IS.append(nums[i])

        return len(IS)
